### PR TITLE
using ThreadLocalPtr to hide ROCKSDB_SUPPORT_THREAD_LOCAL from public…

### PR DESCRIPTION
**Pull request by lightmark**
… headers

https://github.com/facebook/rocksdb/pull/2199 should not reference RocksDB-specific macros (like ROCKSDB_SUPPORT_THREAD_LOCAL in this case) to public headers, `iostats_context.h` and `perf_context.h`. We shouldn't do that because users have to provide these compiler flags when building their binary with RocksDB.

We should hide the thread local global variable inside our implementation and just expose a function api to retrieve these variables. It may break some users for now but good for long term.

make check -j64

### CODE REVIEW

**Comment by igorcanadi** [comment]
Thanks @lightmark , this will fix the problem, although it might be a bit disruptive, since it will cause any code that uses perf_context to stop compiling. To help out with the conversion, I'd add `#define ROCKSDB_PERF_CONTEXT_FUNCTION`, so that people can do
```
#ifdef ROCKSDB_PERF_CONTEXT_FUNCTION
GetPerfContext() something
#else
perf_context something
#endif
```
in their codebases.

An alternative would be to create a build-process-generated rocksdb-defs.h public header file that would be full of  necessary`#defines`.

File: include/rocksdb/perf_context.h [hunk 1]
```diff
@@ -150,21 +150,10 @@ struct PerfContext {
...
-      // Thread local storage on Linux
-      // There is thread_local in C++11
-      #define __thread __declspec(thread)
-    #endif
-    extern __thread PerfContext perf_context;
-  #endif
-#endif
+// Get Thread-local PerfContext object pointer
+// if defined(NPERF_CONTEXT), then the pointer is not thread-local
+PerfContext* GetPerfContext();
```
Comment (siying) [inline include/rocksdb/perf_context.h]: I think the naming convention is `get_perf_context()`.

Comment (lightmark) [inline include/rocksdb/perf_context.h]: how do you think about @igorcanadi 's comment?

Comment (ajkr) [inline include/rocksdb/perf_context.h]: i think do it :), applications should be able to write code compatible across rocksdb minor versions

**Comment by lightmark** [comment]
@igorcanadi What is the benefit of this MACRO over a version guard?

**Comment by igorcanadi** [comment]
> @igorcanadi What is the benefit of this MACRO over a version guard?

That works too. I actually think that the best solution would be to hack the build process to generate rocksdb-defines.h file. This is standard practice: https://cmake.org/cmake/help/v3.0/module/GenerateExportHeader.html

**Comment by lightmark** [comment]
@igorcanadi Let me read. I am not a cmake expert :(

**Pull request closed without merging**